go/merkletree2/tree_test.go

```diff
@@ -746,6 +752,317 @@ func TestNodeEncodingBasic(t *testing.T) {
+	require.Error(t, err)
+	require.Contains(t, err.Error(), "The proof does not have the expected number of root hashes")
+
+	err = verifier.VerifyExtensionProof(NewLoggerContextTodoForTesting(t), &eProof, startSeqno+1, rootHashes[startSeqno+1], endSeqno, rootHashes[endSeqno])
+	require.Error(t, err)
+	require.Contains(t, err.Error(), "The proof does not have the expected number of roots")
+
+	err = verifier.VerifyExtensionProof(NewLoggerContextTodoForTesting(t), &eProof, startSeqno+1, rootHashes[startSeqno], endSeqno, rootHashes[endSeqno+1])
+	require.Error(t, err)
+	require.Contains(t, err.Error(), "The proof does not have the expected number of roots")
+
+	// no previous roots
+	fakeEProof := eProof
+	fakeEProof.PreviousRootsNoSkips = nil
+	err = verifier.VerifyExtensionProof(NewLoggerContextTodoForTesting(t), &fakeEProof, startSeqno, rootHashes[startSeqno], endSeqno, rootHashes[endSeqno])
+	require.Error(t, err)
+	require.Contains(t, err.Error(), "The proof does not have the expected number of roots")
+
+	// no root hashes
+	fakeEProof = eProof
+	fakeEProof.RootHashes = nil
+	err = verifier.VerifyExtensionProof(NewLoggerContextTodoForTesting(t), &fakeEProof, startSeqno, rootHashes[startSeqno], endSeqno, rootHashes[endSeqno])
+	require.Error(t, err)
+	require.Contains(t, err.Error(), "The proof does not have the expected number of root hashes")
+
+	// a root hash has been tampered with
+	fakeEProof = eProof
+	fakeEProof.RootHashes = make([]Hash, len(eProof.RootHashes))
+	copy(fakeEProof.RootHashes, eProof.RootHashes)
+	fakeHashB := sha256.Sum256([]byte("tampered hash!"))
+	fakeHash := Hash(fakeHashB[:])
+	fakeEProof.RootHashes[1] = fakeHash
+	err = verifier.VerifyExtensionProof(NewLoggerContextTodoForTesting(t), &fakeEProof, startSeqno, rootHashes[startSeqno], endSeqno, rootHashes[endSeqno])
+	require.Error(t, err)
+	require.Contains(t, err.Error(), "hash mismatch")
+
+	// no PreviousRootsNoSkips
+	fakeEProof = eProof
+	fakeEProof.PreviousRootsNoSkips = nil
+	err = verifier.VerifyExtensionProof(NewLoggerContextTodoForTesting(t), &fakeEProof, startSeqno, rootHashes[startSeqno], endSeqno, rootHashes[endSeqno])
+	require.Error(t, err)
+	require.Contains(t, err.Error(), "The proof does not have the expected number of roots")
+
+	// a root metadata has been tampered with
+	fakeEProof = eProof
+	fakeEProof.PreviousRootsNoSkips = make([]RootMetadata, len(eProof.PreviousRootsNoSkips))
+	copy(fakeEProof.PreviousRootsNoSkips, eProof.PreviousRootsNoSkips)
+	fakeRoot := RootMetadata{RootVersion: RootVersionTesting, Seqno: Seqno(257), SkipPointersHash: fakeHash}
+	fakeEProof.PreviousRootsNoSkips[1] = fakeRoot
+	err = verifier.VerifyExtensionProof(NewLoggerContextTodoForTesting(t), &fakeEProof, startSeqno, rootHashes[startSeqno], endSeqno, rootHashes[endSeqno])
+	require.Error(t, err)
+	require.Contains(t, err.Error(), "hash mismatch")
```

Comment:
It would be nice to check that this is a different error from the `"hash mismatch"` above, if it's being caught in two different places in the code.